quinn-proto/src/token_reuse_preventer.rs

```diff
@@ -0,0 +1,218 @@
+//! Limiting clients' ability to reuse tokens from NEW_TOKEN frames
+
+use std::{
+    hash::{Hash as _, Hasher as _},
+    mem::{size_of, swap},
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+
+use rustc_hash::{FxHashSet, FxHasher};
+use tracing::warn;
+
+/// Error for when a token may have been reused
+pub struct TokenReuseError;
+
+/// Responsible for limiting clients' ability to reuse tokens from NEW_TOKEN frames
+///
+/// [_RFC 9000 § 8.1.4:_](https://www.rfc-editor.org/rfc/rfc9000.html#section-8.1.4)
+///
+/// > Attackers could replay tokens to use servers as amplifiers in DDoS attacks. To protect
+/// > against such attacks, servers MUST ensure that replay of tokens is prevented or limited.
+/// > Servers SHOULD ensure that tokens sent in Retry packets are only accepted for a short time,
+/// > as they are returned immediately by clients. Tokens that are provided in NEW_TOKEN frames
+/// > (Section 19.7) need to be valid for longer but SHOULD NOT be accepted multiple times.
+/// > Servers are encouraged to allow tokens to be used only once, if possible; tokens MAY include
+/// > additional information about clients to further narrow applicability or reuse.
+pub trait TokenReusePreventer: Send + Sync {
```

Comment:
This name seems kind of awkward. Maybe `TokenLog`?

Reply:
Renaming it sounds ok, maybe `TokenLog` but I feel like that's a significant decrease in accuracy

Reply:
Certainly open to other ideas. Just a bikeshed, anyway.

Reply:
I guess there's `TokenReuseMitigator`. I guess I'm not entirely sure I'm perceiving what you're perceiving about it being awkward--that might be more awkward rather than less.

Reply:
Perhaps `TokenOnceler`, as it makes tokens be used once

![the-onceler.gif](https://github.com/user-attachments/assets/151f0567-3595-4d43-a4f2-a3f8b0715a0b)

No, no.... Not that.

Reply:
Yeah alright I changed it to `TokenLog` it's fine it's shorter.